active/0024-message-validation.md

```diff
@@ -0,0 +1,132 @@
+# Message Filter and Validation
+
+## Changelog
+
+* 2023-10-31: @zmstone Initial draft message validation, filter and transformation.
+* 2022-12-19: @zmstone Limit the scope to filter and validation.
+
+## Abstract
+
+A new feature for EMQX v5 to validate messages.
+
+## Motivation
+
+### Message Validation
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write a rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but they do not interfere MQTT message pub/sub.
+For example, if there is client subscribed to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message filter
+
+Message filtering can be achieved by confugring message validation rule to drop invalid messages.
+
+## Design
+
+This data validation step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being th input.
+
+Each data validation rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+### Message Validation
+
+```
+validation1 {
+    enable = true # or 'false'
+    type = validation
+    description = "drop message if it is not compabitble with my avro schema or if payload.value is less tha 0"
+    topics = "t/#" # or topics = ["t/1", "t/2"]
+    check_mode = any_pass # or all_pass
+    failure_action = disconnect # (disconnect also implies 'dorp') or 'drop' to only drop the message
```

Comment:
Discussion with Ben:

In addition to disconnect & drop message, it should allow the user to specify whether or not to output to the log:
   - Output to system log
   - The log contains basic information such as name, topic, clientid, schema, etc. No payload is needed.
   - Default log level warning

In the future, users should also be allowed to configure log message templates and set log levels.

Users can choose multiple actions, so there should be no drop message on disconnect.

The UI looks like this:
- [x] disconnect
- [ ] drop message 
- [x] output to log

Reply:
This is easy, but in the future it will also need to support user-configurable log output templates.

Reply:
ok, we may add log level as a position in the output template, but the level value itself will still deserve a separate config.

Reply:
Is it possible to add log message templates in version 1?
And can the failure action be configured this way?
```
{
   ...
   failure_action = [
      { "type": "disconnect" },
      { "type": "drop" },
      { 
        "type": "log", 
        "level": "debug", 
        "log_format": "Client ${clientid} - topic ${topic}: message is in the wrong format and the client will be disconnected"
      }
   ]
}
```

Reply:
not quite feasible to add log format here.
what we can do is to add a log `tag` for message validations.
then a separate feature to support dynamic log handler which can be configured to filter certain tags and customize logging fields.